content/blog/2019-11-13-null-pointer-guards.md

```diff
@@ -0,0 +1,220 @@
+`/usr/local/opt/llvm/bin/clang++ -O2 binary-trees.cpp -o binary-trees`
+- *No NPA* is compiling using the null check passes, but without taking into
+consideration which pointers are `PossiblyNull` or `DefinitelyNonNull`. That is,
+we do a null check on every pointer dereference.
+NOTE: Had to go into source code and remove the check for if pointers are `PossiblyNull`
+Compiled with `../tests/compile-with-opt.sh nbody.cpp -O2`
+- *With NPA* is compiling using the null check passes *and* eliding null checks
+if the pointer is `DefinitelyNonNull`.
+Compiled with same instruction as above.
+
+**Runtime Means and Standard Deviations**
+
+|                 |      Baseline      |      No NPA      |     With NPA     |
+|:---------------:|:------------------:|:----------------:|:----------------:|
+|  binary-trees   |  208 ms ± 4.78 ms  | 217 ms ± 12.8 ms | 213 ms ± 6.78 ms |
+|  fannkuch-redux |  222 ms ± 5.02 ms  | 536 ms ± 6.34 ms | 563 ms ± 54.9 ms |
+|  fasta          |  200 ms ± 6.01 ms  | 365 ms ± 7.77 ms | 375 ms ± 28.8 ms |
+|  mandelbrot     |  319 ms ± 5.79 ms  | 294 ms ± 2.52 ms | 298 ms ± 9.76 ms |
+|  n-body         |  327 ms ± 16.7 ms  | 733 ms ± 12.4 ms | 734 ms ± 17.8 ms |
+
+Here we can see that in majority of cases, the baseline performs significantly
+better than the programs with null checks. To me this is a little surprising
+because I figured the branch predictor would always know that the null checks
+don't do anything except return from the null checking function.
+
+It is also interesting to note that the NPA didn't cause any improvement! This
+may be because not that many null pointer checks were removed, perhaps because
+the analysis is too conservative; this would require further investigation.
+This is a little disappointing because I spent a lot of time implementing the
+NPA to reduce the overhead of null checks.
+
+## Hardest Parts to Get Right
+One of the hardest things of this project was the learning curve of LLVM. The
+documentation is fairly good, but there's just not much information overall on
+how to do specific things. For example, I spent a lot of time just figuring out
+how to make a call to `printf` in the IR. For some reason, it wouldn't work
+if `printf` wasn't already used in the file being compiled.
+
+The other hardest thing was doing the null pointer analysis. It was frustrating
+to know that I couldn't check if a pointer aliased nullptr. I wasted a lot of time
+on an incorrect solution that looks as follows:
+> Create a global pointer that is
+> always `nullptr`, and replace all instances of `nullptr` with this global pointer.
+> Then we can use the alias analysis to check what aliases this pointer to see what
+> is potentially null at an instruction. However, by doing this, now all writes to
+> any pointer ends up writing to the location pointed to by the global variable,
+> which is incorrect.
+
+This project gives me a newfound appreciation for compilers writers.
```

Comment:
:tada: :tada: :tada: :cry: :cry: :cry:

It's true; nothing is as easy as it seems!